Add output and logging utils

aedart opened the pull request:
PR adds a few console output utilities. Among them are:

* An `output()` function, that acts as a wrapper for cmake's `message()`.
* Predefined ANSI escape sequences, which can be enabled (_disabled by default when consuming this package_).
* PSR-3 inspired log levels and functions, which respect cmake's current logging level and message modes.

aedart committed:
Add top-level output module (incomplete)

aedart committed:
Add ANSI sub-module

aedart committed:
Include ANSI sub-module

aedart committed:
Use requires_arguments() for assert compare values

aedart committed:
Shorten "escape character" input variable

Also, allow provided code to contain multiple values (list form), due to expected semicolons in code.

aedart committed:
Add tests for ANSI output utils

aedart committed:
Fix RSP_ANSI_PRESET not defined for entire project

This was a HUGE issue to locate - the predefined preset was simply NOT defined, for reasons that escape me. A dump() revealed that despite the include_guard(GLOBAL), the output/ansi was somehow included twice, despite the the guard that should prevent it. This made cmake behave really strange!

aedart committed:
Add demo output function

Useful for a "quick" test / showcase

aedart committed:
Enable ANSI, when project is the root project

aedart committed:
Change release notes

aedart committed:
Add output() function

aedart committed:
Fix ANSI sequence variables only available for current scope

This caused the "include" issue where the RSP ANSI PRESET list was no longer available, at the time when ANSI was enabled, because (in this case) the "rsp/output" was first included inside the test(s) files, meaning that sequence variables were only define within that file's scope!

aedart committed:
Add tests for output()

aedart committed:
Refactor output(), extract message mode handling into macro

Other output util functions will most likely also benefit from this.

aedart committed:
Add logging module (incomplete)

aedart committed:
Preserve message formatting, even when cmake's warn and above takes over

This isn't 100%, but at least the colouring is preserved, and some of the double new lines have been removed. There is still a single double line that I was unable to figure out why it is being added.

aedart committed:
Add RSP_LOG_INDENT

Also, reformatted some of the log components.

aedart committed:
Improve function description

aedart committed:
Change resolve_msg_mode() to accept a default mode

Macro is now slightly more flexible.

aedart committed:
Resolve cmake message mode, acc. to PSR log level

The map should allow developers to customise this in any way they wish.
Also, added output capture logic.

aedart committed:
Fix missing include guard

aedart committed:
Refactor, extract formatting utils into sub-module

aedart committed:
Refactor, extract resolve list separator into macro

aedart committed:
Refactor log(), extract resolve cmake msg mode logic into macro

aedart committed:
Refactor, use resolve_list_separator()

aedart committed:
Fix formatting

aedart committed:
Add RSP_LOG_TIMESTAMP_UTC state

This allows developers to use UTC timestamps, instead of local

aedart committed:
Change list separator handling

This should improve the formatting, slightly... But still isn't what it should be. Damn message() behaves really strange - cmake's "simple markup language" is NOT very well documented.

aedart committed:
Add remaining logging functions from PSR

Need to make a small macro to help forward calls to log(), or the overhead of boilerplate script would be too much...

aedart committed:
Add string contains / not contains asserts

aedart committed:
Add tests for string contains / not contains asserts

aedart committed:
Add examples / ref. for string contains / not contains asserts

aedart committed:
Fix strange name conflict

For reasons that are beyond me, the variable "name" was suddenly shown in a log's context, when developer passed a "name" variable! I think this is because of the macro's variable scope handling.

aedart committed:
Add tests for log functions

aedart committed:
Add docs index page for output module

aedart committed:
Add doc for output ANSI

aedart committed:
Improve examples for text styles

Also, added missing TEXT_BLINK_RESTORE property

aedart committed:
Shorten intro text

aedart committed:
Fix custom styles example

aedart committed:
Add doc for output()

aedart committed:
Fix typo

aedart committed:
Highlight output module

aedart committed:
Add doc index page for logging module

aedart committed:
Add description of the supported log levels

aedart committed:
Improve sentence

aedart committed:
Improve descriptions of the levels and modes

aedart committed:
Fix typo

aedart committed:
Fix typos

aedart committed:
Fix descriptions

aedart committed:
Fix typo

aedart committed:
Add descriptions and examples for each log function

aedart committed:
Change release notes

aedart committed:
Enable ANSI, when project is top-level

kernelguy reviewed (approved):
Perhaps for next update add possibility to control ansi_(en/dis)able via cmake command line option.

aedart committed:
Change root CMakeLists, add RSP_ENABLE_ANSI output option

Project will now enable or disable ANSI output, depending on the option. Defaults to false.

aedart committed:
Change how to enable / disable ANSI output

Now using the newly added RSP_ENABLE_ANSI option.

aedart committed:
Highlight logging module

aedart committed:
Cleanup
